### ISSUE TITLE
Flask-login break IAuthenticator identify interface

### ISSUE THREAD

**Issue by Zharktas** [ckan/ckan#7581]
## CKAN version

2.10.0

## Describe the bug

IAuthenticator identify interface says that extensions should set `g.user` and `g.userobj` https://docs.ckan.org/en/2.10/extensions/plugin-interfaces.html#ckan.plugins.interfaces.IAuthenticator.identify

This works fine on 2.9 and setting those objects lets developer to authenticate users via some other method.

On 2.10, `g.user` is over written with `current_user` from flask-login immediately after the interface calls and the authentication is broken.
https://github.com/ckan/ckan/blob/b2afa91a8e57bdeac874c58b144256732577d9a2/ckan/views/__init__.py#L80-L95

### Steps to reproduce
Implement extension with IAuthenticator identify method and set `g.user` and `g.userobj` in there.
Try to authenticate with it.

### Expected behavior
User should be authenticated with given user object.

**Comment by jonathansberry** [comment]
Yes super interested in seeing this fixed. I've spent a few hours wrestling with this issue already. 

Getting the current_user to change is a little tricky. You can call `toolkit.login_user` but that rewrites the cookie, which I don't want to do. 

My use case, just in case it is helpful:
(a) authenticate the user 
(b) check the user is a sysadmin
(c) optionally identify the sysadmin as a different user for the duration of that request only. But only for that request (i.e. the cookie should be re written). 

At the moment, the only way I've figured so far to set current_user to a desired value for the duration of the request, without rewriting the cookie, is:

```
from flask import current_app
current_app.login_manager._update_request_context_with_user(desired_ckan_user_obj)
```

It feels pretty messy to do this from an extension. I'm also pretty new to flask login, so maybe I'm missing something and there is a better way. 

Interested to follow the outcome of this issue.

**Comment by Zharktas** [comment]
Our hacky solutions for this is using `login_user` and `logout_user` if the user is not identified within identify-function, but that alone kinda breaks api keys. It should be handled by ckan itself and just let extensions set `g.user` and `g.userobj`

**Comment by TomeCirun** [comment]
@Zharktas let me know if this works for you
